### PR TITLE
[filter] Support complex output types for pytorch @open sesame 2/10 13:13

### PR DESCRIPTION
**Pull request by kparichay**
This patch adds support for complex output types for pytorch, where the
complex output is serialized before being passed out to the gstreamer
based element - tensorfilter.

See also: #3620

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc [hunk 1]
```diff
@@ -388,13 +397,65 @@ TorchCore::processIValue (torch::jit::IValue value, GstTensorMemory *output, uns
...
+  if (value.isTensor ()) {
+    if (processIValue (value, output, *idx)) {
+      ml_loge ("Output Tensor Information is not valid");
+      return -2;
+    }
+    (*idx)++;
+  } else if (value.isTuple ()) {
+    auto output_elements = value.toTuple ()->elements ();
+    for (auto element : output_elements) {
+      if (serializeOutput (element.toTensor (), output, idx)) {
```
[the remark this answers is not in the document]
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:434]: This function retains the functionality of the original implementation - https://github.com/nnstreamer/nnstreamer/pull/3653/files/6f11e3d710145f88f36a663a60dbe10a2921c1d1#diff-6f331666401fcd039a035347721e6c97fb15f7fd94c88d6598f37f779c7fa90aL475. So, I am not sure about this.

[the remark this answers is not in the document]
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc]: `.toTensor()` is wrong as you pointed out. Updated.

**Review by anyj0527 — APPROVED** [review]
LGTM 👍. I tested with yolov5s torch model.

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc [hunk 2]
```diff
@@ -310,6 +313,12 @@ TorchCore::validateOutputTensor (const at::Tensor output, unsigned int idx)
   at::Tensor sliced_output = output.slice (0);
   c10::IntArrayRef sliced_output_sizes = sliced_output.sizes ();
 
+  /** if idx is in bounds */
+  if (outputTensorMeta.num_tensors <= idx) {
+    ml_loge ("Invalid output meta: different number of outputs");
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:318]: Please be more specific for error messages. For example, you need to show the desired number of outputs and the actual number of outputs and describe how the user should update their code.

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc [hunk 3]
```diff
@@ -388,13 +397,65 @@ TorchCore::processIValue (torch::jit::IValue value, GstTensorMemory *output, uns
   output_tensor = output_tensor.contiguous ();
 
   /* validate output tensor once */
-  if (configured < outputTensorMeta.num_tensors && validateOutputTensor (output_tensor, idx)) {
+  if (configured < outputTensorMeta.num_tensors
+&& validateOutputTensor (output_tensor, idx)) {
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:401]: Strange indentation...

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc [hunk 4]
```diff
@@ -388,13 +397,65 @@ TorchCore::processIValue (torch::jit::IValue value, GstTensorMemory *output, uns
...
+ * @param[in] value IValue containing the output in tensor form
+ * @param[out] output Output tensor memory
+ * @param[inout] idx index of output
+ * @return 0 if OK. non-zero if error.
+ *         -2 if output tensor validation fails.
+ *         -3 if output is of unsupported format.
+ */
+int
+TorchCore::serializeOutput (
+    const torch::jit::IValue &value, GstTensorMemory *output, unsigned int *idx)
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:422]: Indentation looks strange here, too.

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc [hunk 5]
```diff
@@ -388,13 +397,65 @@ TorchCore::processIValue (torch::jit::IValue value, GstTensorMemory *output, uns
...
+      ml_loge ("Output Tensor Information is not valid");
+      return -2;
+    }
+    (*idx)++;
+  } else if (value.isTuple ()) {
+    auto output_elements = value.toTuple ()->elements ();
+    for (auto element : output_elements) {
+      if (serializeOutput (element, output, idx)) {
+        ml_loge ("Output Tensor Information is not valid");
+        return -2;
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:436]: Please be more specific with error messages here as well. (LINE 427, 435)

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc [hunk 6]
```diff
@@ -388,13 +397,65 @@ TorchCore::processIValue (torch::jit::IValue value, GstTensorMemory *output, uns
...
+    c10::ArrayRef<torch::jit::IValue> output_ref_list = value.toListRef ();
+    std::vector<torch::jit::IValue> output_list (
+        output_ref_list.begin (), output_ref_list.end ());
+#else
+  } else if (value.isGenericList ()) {
+    c10::ArrayRef<torch::jit::IValue> output_list = value.toGenericListRef ();
+#endif
+    for (auto &element : output_list) {
+      if (serializeOutput (element, output, idx)) {
+        ml_loge ("Output Tensor Information is not valid");
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:450]: Need a better error message as well. (e.g., which parameter is invalid & why it's invalid (it shouldn't be NULL, or what has failed, ...) & how to correct it.)

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc [hunk 7]
```diff
@@ -388,13 +397,65 @@ TorchCore::processIValue (torch::jit::IValue value, GstTensorMemory *output, uns
...
+        return -2;
+      }
+    }
+  } else {
+    ml_loge ("Output is not a tensor/tuple/list.");
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:455]: Update the error message here as well.

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc [hunk 8]
```diff
@@ -464,41 +525,10 @@ TorchCore::invoke (const GstTensorMemory *input, GstTensorMemory *output)
...
-      }
-      ++idx;
-    }
-  } else {
-    ml_loge ("Output is not a tensor.");
-    return -3;
-  }
+  unsigned int idx = 0;
+  int retval = serializeOutput (output_value, output, &idx);
+  if (retval)
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:530]: no error messages for this?

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc (hunk 3 above)
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:401]: I am not sure why this clang produces this - fixed manually.

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc (hunk 2 above)
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:318]: Updated the error message to be more specific and describe to user as to how resolve this error.

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc (hunk 5 above)
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:436]: This patch was originally intended to re-format the existing code to support recursive realization.
I will update these error messages to be more descriptive :).

Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:436]: More descriptive messages have been added at the source of the error. 
This location is just passing the error up the call stack and contains index information that has been added to the error message.

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc (hunk 6 above)
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:450]: More descriptive messages have been added at the source of the error.
This location is just passing the error up the call stack and contains index information that has been added to the error message.

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc (hunk 7 above)
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:455]: Updated.

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc (hunk 8 above)
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc]: added.

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc (hunk 4 above)
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:429]: I am not sure what is the desired indentation here as the clang-format recommends this indentation. Please let me know the desired format and I will update at other locations in this file too.

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc [hunk 9]
```diff
@@ -310,19 +313,29 @@ TorchCore::validateOutputTensor (const at::Tensor output, unsigned int idx)
   at::Tensor sliced_output = output.slice (0);
   c10::IntArrayRef sliced_output_sizes = sliced_output.sizes ();
 
+  /** if idx is in bounds */
+  if (outputTensorMeta.num_tensors <= idx) {
+    ml_loge ("Invalid output meta: trying to access index %d with total %d tensors",
+        idx, outputTensorMeta.num_tensors);
+    ml_loge ("Update the number of outputs to >=%d in the model description", idx + 1);
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:320]: Please call ml_loge a single time for a single error. It's a simple stderr output now, but we can add more semantics with a ml_loge call later. (please simply merge the two strings into one.)

Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc]: done.

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc [hunk 10]
```diff
@@ -310,19 +313,31 @@ TorchCore::validateOutputTensor (const at::Tensor output, unsigned int idx)
   at::Tensor sliced_output = output.slice (0);
   c10::IntArrayRef sliced_output_sizes = sliced_output.sizes ();
 
+  /** if idx is in bounds */
+  if (outputTensorMeta.num_tensors <= idx) {
+    ml_loge ("Invalid output meta: trying to access index %d with total %d tensors. "
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc:318]: For users grepping error messages, a string is not supposed to be line-broken. (even in 80col-strict Linux communities). Please keep the string in a single double-quote pair.

**Review by myungjoo — APPROVED** [review]
LGTM except for minor style issues on line-breaking strings. Do not line-break strings.

**Review by jaeyun-jung — APPROVED** [review]

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc (hunk 10 above)
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc]: done. I wonder if there are code search tools to extend searching into strings/messages broken into different lines?

**Comment by kparichay** [comment]
V3:
Merged broken line strings into a single line as requested (for improved grepping).

File: ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc (hunk 10 above)
Comment (myungjoo) [reply ext/nnstreamer/tensor_filter/tensor_filter_pytorch.cc]: We still have a lot of people using grep. :)

**Pull request merged**